File: multi_language_code_generator/implementations/javascript/log_analyzer.js
const fs = require("fs");

const VALID_LEVELS = new Set([
    "DEBUG",
    "INFO",
    "WARNING",
    "ERROR",
    "CRITICAL"
]);

function parseLine(line) {
    const parts = line.split("|", 4).map(part => part.trim());

    if (parts.length !== 4) {
        throw new Error("Malformed log entry");
    }

    const [timestamp, level, source, message] = parts;

    if (!timestamp) {
        throw new Error("Missing timestamp");
    }

    if (!VALID_LEVELS.has(level)) {
        throw new Error("Unknown log level");
    }

    if (!source) {
        throw new Error("Missing source");
    }

    return {
        timestamp,
        level,
        source,
        message
    };
}

function analyze(logs, keywords = [], threshold = 2) {
    if (threshold < 1) {
        throw new Error("Threshold must be at least 1");
    }

    const levelCounts = {};
    const keywordMatches = {};
    const sourceCounts = {};

    for (const keyword of keywords) {
        const normalized = keyword.trim().toLowerCase();

        if (normalized) {
            keywordMatches[normalized] = 0;
        }
    }

    let totalEntries = 0;

    for (const line of logs) {
        let entry;

        try {
            entry = parseLine(line);
        } catch (error) {
            continue;
        }

        totalEntries++;

        const level = entry.level;
        levelCounts[level] = (levelCounts[level] || 0) + 1;

        const source = entry.source;
        sourceCounts[source] = (sourceCounts[source] || 0) + 1;

        const message = entry.message.toLowerCase();

        for (const keyword of Object.keys(keywordMatches)) {
            let position = 0;

            while ((position = message.indexOf(keyword, position)) !== -1) {
                keywordMatches[keyword]++;
                position += keyword.length;
            }
        }
    }

    const repeatedSources = Object.keys(sourceCounts)
        .filter(source => sourceCounts[source] >= threshold)
        .sort();

    return {
        total_entries: totalEntries,
        level_counts: levelCounts,
        keyword_matches: keywordMatches,
        repeated_sources: repeatedSources
    };
}

if (require.main === module) {
    const input = fs.readFileSync(0, "utf8");
    const data = JSON.parse(input);

    const result = analyze(
        data.logs || [],
        data.keywords || [],
        data.threshold || 2
    );

    console.log(JSON.stringify(result, null, 2));
}

module.exports = {
    parseLine,
    analyze
};

File: multi_language_code_generator/implementations/python/log_analyzer.py
import json
import sys


VALID_LEVELS = {
    "DEBUG",
    "INFO",
    "WARNING",
    "ERROR",
    "CRITICAL",
}


def parse_line(line):
    parts = [part.strip() for part in line.split("|", 3)]

    if len(parts) != 4:
        raise ValueError("Malformed log entry")

    timestamp, level, source, message = parts

    if not timestamp:
        raise ValueError("Missing timestamp")

    if level not in VALID_LEVELS:
        raise ValueError("Unknown log level")

    if not source:
        raise ValueError("Missing source")

    return {
        "timestamp": timestamp,
        "level": level,
        "source": source,
        "message": message,
    }


def analyze(logs, keywords=None, threshold=2):
    if keywords is None:
        keywords = []

    if threshold < 1:
        raise ValueError("Threshold must be at least 1")

    level_counts = {}
    keyword_matches = {}
    source_counts = {}

    for keyword in keywords:
        keyword = keyword.strip().lower()
        if keyword:
            keyword_matches.setdefault(keyword, 0)

    total_entries = 0

    for line in logs:
        try:
            entry = parse_line(line)
        except ValueError:
            continue

        total_entries += 1

        level = entry["level"]
        level_counts[level] = level_counts.get(level, 0) + 1

        source = entry["source"]
        source_counts[source] = source_counts.get(source, 0) + 1

        message = entry["message"].lower()

        for keyword in keyword_matches:
            keyword_matches[keyword] += message.count(keyword)

    repeated_sources = sorted(
        source
        for source, count in source_counts.items()
        if count >= threshold
    )

    return {
        "total_entries": total_entries,
        "level_counts": level_counts,
        "keyword_matches": keyword_matches,
        "repeated_sources": repeated_sources,
    }


if __name__ == "__main__":
    data = json.load(sys.stdin)

    result = analyze(
        data.get("logs", []),
        data.get("keywords", []),
        data.get("threshold", 2),
    )

    print(json.dumps(result, indent=2))